### PR TITLE
[ownership] Extract out SILOwnershipKind from ValueOwnershipKind into its own type and rename Invalid -> Any.

### PR DESCRIPTION
**Pull request by gottesmm**
This makes it easier to understand conceptually why a ValueOwnershipKind with
Any ownership is invalid and also allowed me to explicitly document the lattice
that relates ownership constraints/value ownership kinds.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@atrick I also put in some ascii art that shows the lattice: https://github.com/apple/swift/pull/34669/files#diff-7f261169f338bde9847b6fc72d36c2657ab1e46e8dfd4db1bd834e8657edba13R91

File: include/swift/SIL/SILValue.h [hunk 1]
```diff
@@ -195,9 +285,9 @@ struct ValueOwnershipKind {
   /// that the two ownership kinds are "compatibile".
   ///
   /// The reason why we do not compare directy is to allow for
-  /// ValueOwnershipKind::None to merge into other forms of ValueOwnershipKind.
+  /// OwnershipKind::None to merge into other forms of ValueOwnershipKind.
   bool isCompatibleWith(ValueOwnershipKind other) const {
-    return bool(merge(other));
+    return *this == merge(other);
```
Comment (atrick) [inline include/swift/SIL/SILValue.h:290]: If `isCompatibleWith` is supposed to mean "a merge produces a valid value", then I think you need to test that directly, like the previous implementation:

`bool(merge(other))`

The implementation here is asymmetric.

Comment (gottesmm) [reply include/swift/SIL/SILValue.h:290]: Yea. I hit this. It was a mistake.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Review by atrick — APPROVED** [review]
Perfect, except for `isCompatibleWith`. Although I might have missed other functional changes...

**Comment by gottesmm** [comment]
That was the only change I made. Everything else is the same.

**Comment by gottesmm** [comment]
@swift-ci clean smoke test OS X platform

**Pull request merged**